runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectJobBundleFactory.java

```diff
@@ -0,0 +1,171 @@
+        SdkHarnessClient.usingFnApiClient(
+            remoteEnv.getInstructionRequestHandler(), dataService.getService());
+    ExecutableProcessBundleDescriptor descriptor;
+    try {
+      descriptor =
+          ProcessBundleDescriptors.fromExecutableStage(
+              Long.toString(idgen.getAndIncrement()), stage, dataService.getApiServiceDescriptor());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    RemoteInputDestination<? super WindowedValue<?>> destination =
+        descriptor.getRemoteInputDestination();
+    SdkHarnessClient.BundleProcessor<T> bundleProcessor =
+        sdkHarnessClient.getProcessor(
+            descriptor.getProcessBundleDescriptor(),
+            (RemoteInputDestination<WindowedValue<T>>) (RemoteInputDestination) destination,
+            stateService.getService());
+    return new DirectStageBundleFactory<>(descriptor, bundleProcessor);
+  }
+
+  @Override
+  public void close() throws Exception {
+    Exception thrown = null;
+    for (RemoteEnvironment remoteEnvironment : environments.values()) {
+      try {
+        remoteEnvironment.close();
+      } catch (Exception e) {
+        if (thrown == null) {
+          thrown = e;
+        } else {
+          thrown.addSuppressed(e);
+        }
+      }
+    }
+    if (thrown != null) {
+      throw thrown;
+    }
+  }
+
+  private static class DirectStageBundleFactory<T> implements StageBundleFactory<T> {
+    private final ExecutableProcessBundleDescriptor descriptor;
+    private final SdkHarnessClient.BundleProcessor<T> processor;
+
+    private DirectStageBundleFactory(
+        ExecutableProcessBundleDescriptor descriptor,
+        SdkHarnessClient.BundleProcessor<T> processor) {
+      this.descriptor = descriptor;
+      this.processor = processor;
+    }
+
+    @Override
+    public RemoteBundle<T> getBundle(
+        OutputReceiverFactory outputReceiverFactory, StateRequestHandler stateRequestHandler)
+        throws Exception {
+      Map<Target, RemoteOutputReceiver<?>> outputReceivers = new HashMap<>();
+      for (Map.Entry<Target, Coder<WindowedValue<?>>> targetCoders :
+          descriptor.getOutputTargetCoders().entrySet()) {
+        String bundleOutputPCollection =
+            Iterables.getOnlyElement(
+                descriptor
+                    .getProcessBundleDescriptor()
+                    .getTransformsOrThrow(targetCoders.getKey().getPrimitiveTransformReference())
+                    .getInputsMap()
+                    .values());
+        FnDataReceiver<WindowedValue<?>> outputPCollection =
+            outputReceiverFactory.create(bundleOutputPCollection);
+        outputReceivers.put(
+            targetCoders.getKey(),
+            RemoteOutputReceiver.of(targetCoders.getValue(), outputPCollection));
+      }
+      ActiveBundle<T> bundle = processor.newBundle(outputReceivers, stateRequestHandler);
+      MoreFutures.get(processor.getRegistrationFuture());
```

Comment:
I don't believe we need to expose `getRegistrationFuture` and should be able to remove the method as it is internally handled within the `ActiveBundle`.

Reply:
rm'd. It's not currently handled in `ActiveBundle`, but that's the correct place for it.